File: coding/python/02_data_structures/linked_list.py
"""Singly linked list implemented from scratch."""


class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = node

    def prepend(self, value):
        node = Node(value)
        node.next = self.head
        self.head = node

    def delete(self, value):
        if not self.head:
            return False
        if self.head.value == value:
            self.head = self.head.next
            return True
        current = self.head
        while current.next:
            if current.next.value == value:
                current.next = current.next.next
                return True
            current = current.next
        return False

    def find(self, value):
        current = self.head
        while current:
            if current.value == value:
                return True
            current = current.next
        return False

    def reverse(self):
        prev = None
        current = self.head
        while current:
            nxt = current.next
            current.next = prev
            prev = current
            current = nxt
        self.head = prev

    def to_list(self):
        result = []
        current = self.head
        while current:
            result.append(current.value)
            current = current.next
        return result

    def __repr__(self):
        return " -> ".join(str(v) for v in self.to_list())


def main():
    ll = LinkedList()
    for v in [1, 2, 3, 4, 5]:
        ll.append(v)
    print("list:", ll)

    ll.prepend(0)
    print("after prepend(0):", ll)

    ll.delete(3)
    print("after delete(3):", ll)

    print("find(4):", ll.find(4))
    print("find(99):", ll.find(99))

    ll.reverse()
    print("reversed:", ll)


if __name__ == "__main__":
    main()

File: coding/c/02_data_structures/linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node *next;
} Node;

typedef struct {
    Node *head;
} LinkedList;

Node *node_create(int value) {
    Node *node = malloc(sizeof(Node));
    if (!node) {
        fprintf(stderr, "malloc failed\n");
        exit(1);
    }
    node->value = value;
    node->next = NULL;
    return node;
}

void list_append(LinkedList *list, int value) {
    Node *node = node_create(value);
    if (!list->head) {
        list->head = node;
        return;
    }
    Node *current = list->head;
    while (current->next) {
        current = current->next;
    }
    current->next = node;
}

int list_delete(LinkedList *list, int value) {
    Node *current = list->head;
    Node *prev = NULL;

    while (current) {
        if (current->value == value) {
            if (prev) {
                prev->next = current->next;
            } else {
                list->head = current->next;
            }
            free(current);
            return 1;
        }
        prev = current;
        current = current->next;
    }
    return 0;
}

void list_print(const LinkedList *list) {
    Node *current = list->head;
    while (current) {
        printf("%d", current->value);
        if (current->next) {
            printf(" -> ");
        }
        current = current->next;
    }
    printf("\n");
}

void list_free(LinkedList *list) {
    Node *current = list->head;
    while (current) {
        Node *next = current->next;
        free(current);
        current = next;
    }
    list->head = NULL;
}

int main(void) {
    LinkedList list = {NULL};

    for (int i = 1; i <= 5; i++) {
        list_append(&list, i);
    }
    printf("list: ");
    list_print(&list);

    list_delete(&list, 3);
    printf("after delete(3): ");
    list_print(&list);

    list_free(&list);
    printf("list freed\n");

    return 0;
}
